common/src/main/java/org/opensearch/sdk/client/LocalClusterIndicesClient.java

```diff
@@ -189,7 +197,13 @@ public CompletionStage<DeleteDataObjectResponse> deleteDataObjectAsync(DeleteDat
     public CompletionStage<SearchDataObjectResponse> searchDataObjectAsync(SearchDataObjectRequest request, Executor executor) {
         return CompletableFuture.supplyAsync(() -> AccessController.doPrivileged((PrivilegedAction<SearchDataObjectResponse>) () -> {
             log.info("Searching {}", Arrays.toString(request.indices()));
-            SearchResponse searchResponse = client.search(new SearchRequest(request.indices(), request.searchSourceBuilder())).actionGet();
+            SearchSourceBuilder searchSource = request.searchSourceBuilder();
+            QueryBuilder existingQuery = searchSource.query();
+            String tenantId = request.tenantId() != null ? request.tenantId() : DEFAULT_TENANT;
```

Comment:
In case multi tenancy is disabled tenant ID field will be empty. Searching for documents with `DEFAULT_TENANT` will return incorrect result.

Reply:
Sorry, this is extremely poorly documented.  As you can probably see, Integ tests are failing precisely because of this assumption.

But how, in the `common` module, is it even possible to know if multi-tenancy is disabled?  I moved this client here by request, but that makes it impossible to know whether multi-tenancy is enabled.

Tell me what this line of code should read.   I know it's wrong.  I don't know how to make it right.

Reply:
To be more clear, please answer:
1. When is tenant Id `DEFAULT_TENANT_ID` and when is it null?  Nobody has bothered to document the conventions here.   Why do we need a specific non-null string?
2. Why are we choosing the string "DEFAULT_TENANT_ID" for this value?   Similar to my objection to using the field name/hash key "tenant_id" we should not use any common string that anyone might accidentally use.  Can a user overwrite the `tenant_id` field with a value of `"DEFAULT_TENANT_ID"` and gain access to all data in the cluster?  If not, where is that documented (see #1).  Let's align on these constants, I feel that I have made multiple comments on other PRs that have not led to any discussion or final conclusion.  
3. How is it possible in the `common` module for this code to know the value in the `plugin` module is; specifically whether multi-tenancy is enabled?  What happens when we move this client code to another repository?  How is this external dependency going to know whether multi-tenancy is enabled?    Should it be assumed that if you're using any of the SDK client implementations you have to use multi-tenancy?

I consider this PR and honestly a lot of work on this dev branch blocked until we align on question 3 at least.

Flipping this to draft mode.  @dhrubo-os @arjunkumargiri please comment.

Reply:
@dbwiddis please let me know how can I improve documentation [here](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java#L61-L63)

During index searching, can't we search tenant_id =  null (if tenant id isn't provided)? Why do we need any default value here?


>How is it possible in the common module for this code to know the value in the plugin module is; specifically whether multi-tenancy is enabled? What happens when we move this client code to another repository? How is this external dependency going to know whether multi-tenancy is enabled? Should it be assumed that if you're using any of the SDK client implementations you have to use multi-tenancy?


Yeah, I was having the same issue, then I ended up doing this: https://github.com/opensearch-project/ml-commons/pull/2792/files

In `MLAgentExecutor.java`, which stays in the Algorithm module I needed to have the dynamic multi-tenancy value.

Reply:
> During index searching, can't we search tenant_id = null (if tenant id isn't provided)? 

The filter by term query doesn’t accept null.  We can easily skip adding the filter if we know we’re not multi tenant.  But how do we know that?

and how do we protect against accidentally leaving it null and exposing other tenant information?

> please let me know how can I improve documentation [here](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java#L61-L63)

that just explains why we substitute a string but not how we treat null throughout the end to end. 

Does null always mean multi tenancy is not enabled?  It should but that needs validation somewhere.

Reply:
Yeah for multi-tenancy we are expecting `tenant_id` not to be null. 
For rest layer, check is [here](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/plugin/src/main/java/org/opensearch/ml/rest/AbstractMLSearchAction.java#L60)

For transport layers checks are:

1. [agent search](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/plugin/src/main/java/org/opensearch/ml/action/agents/TransportSearchAgentAction.java#L61-L63)
2. [connector search](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/plugin/src/main/java/org/opensearch/ml/action/connector/SearchConnectorTransportAction.java#L82-L84)
3. [model group search](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/plugin/src/main/java/org/opensearch/ml/action/model_group/SearchModelGroupTransportAction.java#L72-L74)
4. [model search](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/plugin/src/main/java/org/opensearch/ml/action/models/SearchModelTransportAction.java#L49-L51)
5. [task search](https://github.com/opensearch-project/ml-commons/blob/feature/multi_tenancy/plugin/src/main/java/org/opensearch/ml/action/tasks/SearchTaskTransportAction.java#L57-L59)


Please let me know where else we should put this validation?


>The filter by term query doesn’t accept null. We can easily skip adding the filter if we know we’re not multi tenant. But how do we know that?

We already have validation in rest and transport layer for tenant id check. Do we still need to validate tenant ID in SDKClient level? IF that's not necessary, we can skip to add tenant id in the search if tenant Id is null?

Reply:
>that just explains why we substitute a string but not how we treat null throughout the end to end.

And where exactly do you suggest to put this documentation?

Reply:
> We already have validation in rest and transport layer for tenant id check.

Lots of individual checks on individual REST and Transport actions don't give me a warm fuzzy.

> And where exactly do you suggest to put this documentation?

I don't know.  As you know, this feature branch is a collaboration primarily of three individuals and theres no documentation anywhere in code/ javadoc/readme/RFC issue/etc. that tells me how to handle Tenant ID consistently.

I recognize we're rapidly iterating on a feature branch and that's totally fine until now when we're talking about migrating to main, and I'm raising my hand with a big red _"the worst possible thing we can do right now is expose one tenant's data to another tenant"_.  

I really do not feel comfortable that we have talked about what is expected here, and what safeguards we have in place to prevent it.  The fact that integ tests are currently failing on this PR is demonstration of the fact that I do not yet understand what the code expects, how we are going to remain compatible with BWC when existing indices do not have tenant ID in their documents, and how we are going to ensure going forward that tenant ID is always available and any attempt by a user to manipulate it should result in failure.   Currently it seems we're relying on a UUID but it seems that spamming a search with a brute force attack of random UUIDs will eventually uncover some random user's data.  Is UUID sufficient?

People (especially me) make mistakes, and failing to insert the REST or Transport validation checks that are apparently already there may happen without at least a test somewhere proving that they work (and perhaps that's my job to write such a test).  I honestly don't understand enough right now about what is expected to write that test and I'm calling that out.

Reply:
> Currently it seems we're relying on a UUID but it seems that spamming a search with a brute force attack of random UUIDs will eventually uncover some random user's data. Is UUID sufficient?

We should not be relying solely on UUIDs as tenant ID to prevent unauthorized access to other tenants' data.

https://www.rfc-editor.org/rfc/rfc4122#section-6
> Do not assume that UUIDs are hard to guess; they should not be used as security capabilities (identifiers whose mere possession grants access), for example.  A predictable random number source will exacerbate the situation.

Reply:
> Yeah, I was having the same issue, then I ended up doing this: https://github.com/opensearch-project/ml-commons/pull/2792/files
> 
> In `MLAgentExecutor.java`, which stays in the Algorithm module I needed to have the dynamic multi-tenancy value.

That's a great workaround for local files but it doesn't help in the case of SdkClient when we are planning on moving it out of this repository eventually.  It won't have access to that interface.  

That interface belongs in the SDKClient package(s) somewhere.

Reply:
> We should not be relying solely on UUIDs as tenant ID to prevent unauthorized access to other tenants' data.

@dbwiddis Do you mean: A cluster has authN/authZ and multi-tenancy is enabled, tenantA's can access the cluster with their credentials but they tempered the tenantId to tenantB's and accessed to tenantB's data?

Reply:
> That interface belongs in the SDKClient package(s) somewhere.

Can you give an example why this belongs to different package would be an issue? My assumption is there'll be a `tenantId` parameter in the interface and the value will be passed from upper layer, if it's null skip the tenantId filter, otherwise add it.

Reply:
> Do you mean: A cluster has authN/authZ and multi-tenancy is enabled

@zane-neo I had assumed that we were producing the tenant ID similarly to how we were producing document ID.  We are not; it is being supplied by the tenant. I no longer have this concern.

Reply:
> Can you give an example why this belongs to different package would be an issue?

@zane-neo the SdkClient and its implementations are not meant to permanently reside in ML Commons.  Ideally they will be in another dependency entirely, possibly eventually core or a module, possibly another plugin (similar to how skills is separated).

There needs to be a way to have a plugin supply this client the needed settings updater from the plugin module without relying on this being present in ML Commons.

Reply:
> Ideally they will be in another dependency entirely, possibly eventually core or a module, possibly another plugin (similar to how skills is separated).

So this is a highly uncertainty, currently it seems fine to implement `skip if null else filter` now, it's not a hard blocker. If there's any early stage decision how this client will be separated, then that'll be a better place to discuss this concern.

Reply:
> currently it seems fine to implement skip if null else filter now

That's essentially what I've done already, skipping the filter if multi-tenancy isn't enabled -- but given how tightly this particular setting is with the clients it just makes sense to eventually move it (a later PR is fine) alongside the clients implementing it.